### ISSUE TITLE
Lint rules for PEP 723 inline script metadata content

### ISSUE THREAD

**Issue by janosh** [astral-sh/ruff#24417]
## Summary

PEP 723 inline script metadata is widely adopted (`uv`, `hatch`, `pdm`, `pipx`, `pip 26+`), but no linter validates the **content** of these blocks. This is distinct from #10457 (ruff reading metadata to configure its own rules). This proposal is about catching bugs and enforcing consistency **within** the metadata block itself.

Since `# /// script` blocks are embedded in comments, typos and structural errors can go unnoticed until runtime, or be silently ignored.

### Tier 1: catches real bugs / spec violations

**1. Duplicate script metadata blocks** (error)

The [spec](https://packaging.python.org/en/latest/specifications/inline-script-metadata/) states: *"When there are multiple comment blocks of the same TYPE defined, tools MUST produce an error."* In practice, `uv` silently uses only the first block and ignores the second, so the bug is latent, a bit harder to catch.

```python
# /// script
# dependencies = ["requests"]
# ///

# ... 200 lines later ...

# /// script
# dependencies = ["flask"]  # silently ignored by uv
# ///
```

**2. Invalid TOML in metadata block** (error)

The block content must be valid TOML. Invalid TOML causes runtime errors in `uv` but could be caught earlier at lint time.

```python
# /// script
# dependencies = [
#   "requests"   # missing comma; uv: "missing comma between array elements"
#   "flask"
# ]
# ///
```

**3. Invalid PEP 508 dependency specifiers** (error)

The spec says each dependency entry *"MUST be a valid PEP 508 dependency"*. Ruff already has a PEP 508 parser for `I002`.

```python
# /// script
# dependencies = [
#   "requests>=2.0<=3.0",  # invalid; uv: "Trailing `<= 3.0` is not allowed"
# ]
# ///
```

**4. Invalid PEP 440 `requires-python` version specifier** (error)

A bare version without an operator is a common mistake. `uv` helpfully suggests `==3.11` but this could be caught at lint time.

```python
# /// script
# requires-python = "3.11"  # invalid; uv: "Did you mean `==3.11`?"
# ///
```

**5. Duplicate dependencies** (warning, fixable)

```python
# /// script
# dependencies = [
#   "requests",
#   "flask",
#   "requests",  # duplicate
# ]
# ///
```

**6. Unsupported top-level fields** (warning)

Only `dependencies`, `requires-python`, and `[tool]` are valid top-level entries in a `script` block.

```python
# /// script
# name = "my-script"          # not a valid field per spec
# requires-python = ">=3.11"
# ///
```

### Tier 2: code quality / consistency (fixable)

**7. Unsorted dependencies** (warning, fixable)

Alphabetically sorted dependency lists produce cleaner diffs and reduce merge conflicts. Analogous to `I001` for imports and `RUF022` for `__all__`.

```python
# /// script
# dependencies = [
#   "requests",  # should come after numpy
#   "numpy",
#   "flask",     # should come first
# ]
# ///
```

**8. Dependency name normalization** (warning, fixable)

[PEP 503](https://peps.python.org/pep-0503/#normalized-names) specifies canonical package names (lowercase, hyphens normalized). While dependency specifiers are case-insensitive, normalizing improves readability and makes deduplication straightforward.

```python
# /// script
# dependencies = [
#   "Flask",        # could be "flask"
#   "Scikit-Learn", # could be "scikit-learn"
# ]
# ///
```

### Tier 3: best practices

**9. Unused dependencies** (warning)

A dependency is listed but never imported in the script. This overlaps with #10015 (deptry-style checks) but is simpler for single-file scripts: no `sys.path` or environment introspection needed, just cross-reference `dependencies` with `import` statements in the same file.

```python
# /// script
# dependencies = [
#   "requests",
#   "flask",      # never imported
# ]
# ///

import requests
```

**10. Malformed block delimiters** (warning)

Near-miss delimiters are silently ignored. The block isn't parsed, so dependencies aren't installed. No error is raised, which can be confusing to debug.

```python
# ///script         # missing space before "script", silently ignored
# requires-python = ">=3.11"
# dependencies = ["requests"]
# ///
```

The script runs but `requests` is not installed because `uv` doesn't recognize the block.

**11. Missing `requires-python`** (suggestion)

Good practice to specify the Python version range since `uv` uses it to select the interpreter.

## Context

- PEP 723 is Final status, adopted by `uv`, `hatch`, `pdm`, `pipx`, `pip`
- #10457 tracks ruff *consuming* metadata (e.g. respecting `requires-python` for import sorting), orthogonal to this
- #10455 (ERA001 false positives in PEP 723 blocks) was already fixed
- #10015 tracks general unused/missing dependency detection; rule 9 above is the simpler single-file-script subset
- Rules 1-4 enforce MUST-level requirements from the spec
- All behaviors above verified against `uv 0.7.12` on 2026-04-05

**Comment by MichaReiser** [comment]
Thanks for the deailed write up. @dylwil3 might soon add PEP 723 configuration support and that should enable us to also apply the existing pyproject.toml lint rules. 

The specific rules seem to be more around pyproject.toml best practices rather than being specific to script blocks. That's why I'm inclined to repurpose this issue around enabling the pyproject.toml lints for script blocks.

**Comment by janosh** [comment]
thanks for the quick reply! extending RUF200 to script blocks would cover rules 2-4 and 6 (TOML validity, PEP 508/440 specifiers, unsupported fields). i think rules 1, 5, 7-10 are new checks that don't exist for `pyproject.toml` yet (unless i missed it). duplicate blocks, duplicate/unsorted deps, name normalization, unused deps, malformed delimiters would be very nice to have. so roughly half of this proposal is "apply existing validation to a inline script metadata" and the other half is new lints which could also benefit `pyproject.toml`. should i spit the latter into a new issue?

**Comment by MichaReiser** [comment]
Splitting the issue makes sense to me, and we can keep this one to enable pyproject toml linting for script blocks.

Regarding 1, we can generalize this one to catch misplaced script blocks (that won't be recognized as such).
